### PR TITLE
798 fix order fields

### LINKED ISSUE #798: orderFields not working
I want to order the fileds in form, but orderFields not work.

`        // ------ CRUD FIELDS
        $this->crud->orderFields(['name', 'parent_id', 'lft'], 'update/create/both');`

In the form parent_id remaining the first field.
Why?

### PR DESCRIPTION
**Pull request by tumf87**
Fixes #798.

@tabacitu, please notice that, as part of this PR, I've deprecated the unimplemented `setFieldOrder` and `setFieldsOrder`.

Also, since I was already working on the class, I've also fixed the `decodeJsonCastedAttributes` catch clause not declaring the Exception class correctly.

### CODE REVIEW

File: src/PanelTraits/Fields.php [hunk 1]
```diff
@@ -11,60 +11,55 @@
...
+     * @param string|array $field The new field.
+     * @param string $form The CRUD form. Can be 'create', 'update' or 'both'. Default is 'both'.
      */
     public function addField($field, $form = 'both')
     {
         // if the field_definition_array array is a string, it means the programmer was lazy and has only passed the name
         // set some default values, so the field will still work
         if (is_string($field)) {
-            $complete_field_array['name'] = $field;
+            $completeFieldsArray['name'] = $field;
```
Comment (lloy0076) [inline src/PanelTraits/Fields.php]: Is that the field definition array (i.e. is that variable oddly named)?

File: src/PanelTraits/Fields.php [hunk 2]
```diff
@@ -263,13 +216,92 @@ public function getCurrentFields()
         return $this->getCreateFields();
     }
 
-    // ------------
-    // TONE FUNCTIONS - UNDOCUMENTED, UNTESTED, SOME MAY BE USED
-    // ------------
-    // TODO: check them
+    /**
+     * Order the CRUD fields in the given form. If certain fields are missing from the given order array, they will be
```
Comment (lloy0076) [inline src/PanelTraits/Fields.php]: That second sentence probably should be in its own paragraph.

File: src/PanelTraits/Fields.php [hunk 3]
```diff
@@ -263,13 +216,92 @@ public function getCurrentFields()
...
+     */
+    public function orderFields($order, $form = 'both')
+    {
+        $this->transformFields($form, function ($fields) use ($order) {
+            return $this->getOrderedFields($fields, $order);
+        });
+    }
+
+    /**
+     * Apply the given order to the fields and return the new array.
```
Comment (lloy0076) [inline src/PanelTraits/Fields.php]: The method is name `getOrderedField` - should this say "apply" then?

**Comment by tumf87** [comment]
@lloy0076, like the column order PR, the method names were kept for backwards compatibility reasons.

File: src/PanelTraits/Fields.php (hunk 1 above)
Comment (tumf87) [reply src/PanelTraits/Fields.php]: Yeah. I just changed the casing. Any suggestions for a better name?

Comment (lloy0076) [reply src/PanelTraits/Fields.php]: `$allTheFields` (does it need the hungarian notiation)?
`$fieldsInformation`

**Comment by tabacitu** [comment]
Thank you @tumf87 ! Merged!

**Pull request merged**